### ISSUE TITLE
Memory leak in mat-drawer and mat-sidenav

### ISSUE THREAD

**Issue by tran4o** [angular/components#11335]
#### Bug, feature request, or proposal:

Memory leak in mat-drawer, mat-sidenav after destroy

#### What is the expected behavior?

Full cleanup

#### What is the current behavior?
Transition engine not freed-up 

#### What are the steps to reproduce?

Simple show / hide with recreation (*ngIf)

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?

Angular 6.0.0 Material 6.0.2

#### Is there anything else we should know?

Dirty workaround would be a nav-patch directive :
 
<mat-drawer nav-patch  ..

@Directive({
    selector: '[nav-patch]',
    outputs : []
})
export class NavPatch {
     constructor(private eref: ElementRef,private renderer : Renderer) {         
        var t : any = renderer;
        while (t.delegate && (!t.engine || !t.engine._transitionEngine)) t=t.delegate;
        this._transitionEngine = t.engine._transitionEngine;
     }  
     ngOnDestroy() {     
        // PATCH : CLEAR REFERENCES
        this._transitionEngine.statesByElement.delete(this.eref.nativeElement);
        this._transitionEngine.playersByElement.delete(this.eref.nativeElement);
    }
     private _transitionEngine;
}

**Comment by sheiidan** [comment]
I took a look in `MatDrawer` component. It has 2 cases where it subscribe to events without unsubscribe them on `ngOnDestroy` - this looks like a leak to me. 
I patched the code locally and I added unsubscription from the events and the leak is gone.
Here's one of the leaks:
https://github.com/angular/material2/blob/master/src/lib/sidenav/drawer.ts

![image](https://user-images.githubusercontent.com/28801860/49036552-ab9ce080-f1c0-11e8-82a4-228704e88b1b.png)
